### DISCUSSION TITLE
Feature Request: LLHLS Session Timeout

### DISCUSSION THREAD

**Discussion by bchah** [OvenMediaLabs/OvenMediaEngine#1313] [Ideas] [answered]
In #1304 we discussed the issue where HTTP/2.0 connections do not get closed until the browser window is closed, even when the page is reloaded. The result is that OME will hold many connections even if it is only using one of them, and the viewer statistics will be inaccurate.

I know we cannot change HTTP/2.0, but maybe it's possible to solve the problem with another approach...

Would it be possible to add a "Timeout" configuration value to the LLHLS publisher? The purpose of the timeout would be to observe if no data has flowed to that session in the last X seconds, then the session would be destroyed. This would in theory keep the viewer count accurate. I hope you can consider adding this because the inaccurate viewer count is a security concern for users of LLHLS 🥹

**Comment by getroot** [comment, marked as answer]
I think my explanation was misrepresented. OvenMediaEngine will of course close the connection after 30 seconds if no data is flowing on the TCP connection. It is not true that you said that OME holds many connections. The fact that there are multiple sessions is just a statistical number, and there is only one actual TCP connection.

![image](https://github.com/AirenSoft/OvenMediaEngine/assets/17779085/346a7762-6721-4eaf-9d30-4b7e6d9d6a8c)


In detail, it is as follows.

As shown in the figure, the HTTP/2 browser uses only one connection in the entire PC when connecting to one server.

In other words, even if you run 3 browsers on your PC and play 3 videos with OvenPlayer, only one TCP connection is connected to OME.

This makes it difficult for OvenMediaEngine to calculate "concurrent viewers".

There may be several strategies.

1. Calculate concurrent viewers only with the number of TCP connections
   > This is displayed as 1 simultaneous viewer even when playing n streams on one PC. Early on, we had this strategy.

2. The master playlist (default: llhls.m3u8) is taken only once when the player starts, so 1 concurrent view count is added each time the master playlist is downloaded > This has the following weaknesses.

Let's assume the following.
(1) Launch 3 players on one PC and start playback
(2) Quit 2 players and continue playing with only 1

Can OvenMediaEngine recognize that two players are closed at this time?

This is very difficult. Because it is related to the HLS protocol. There is no state called "end" in the HLS protocol. Because the client just downloads the chunklist and segments. Two players are closed, but one player is still downloading chunks. So OvenMediaEngine doesn't know that both players are closed.

So I put an algorithm that determines that all three sessions related to it are closed when the connection is disconnected or when no data flows through the connection.

Please let me know if you have any better algorithm ideas for calculating statistics.

**Reply by bchah** [reply]
Thank you @getroot this is an amazing explanation and makes a lot of sense.

With my current understanding, we have this info available:

- Each user only creates one connection at a time
- Each stream creates a master playlist that is called once per lifecycle of either the connection or the stream
- Each playback session requests chunks from the master playlist, all served on one connection

I only see one missing piece to get our answer: can OME measure the outgoing chunks on a per-playlist basis? 

For example, if I request chunks from Stream1 and Stream2 can we measure the output of data for each stream? The idea is that if we can calculate how much data a connection _should_ be receiving per stream (based on the bitrate of the stream * the number of times the master playlist has been requested in that session lifecycle), we could tell which sessions are active by checking the data rate of chunks for _that_ stream being requested on _that_ connection.

If you can please confirm the missing piece (if we can measure data output per-stream, per-connection) I will deliver a more complete proposal for this idea 😁

**Reply by getroot** [reply]
Considering ABR, it is very difficult to calculate the number of concurrent views with output bandwidth. Also, the number of concurrent viewers will go up and down very quickly as it is very dependent on network speed and jitter etc.

**Reply by bchah** [reply]
How about number of chunks requested? Could we validate the session count on each connection by checking the number of chunks they are requesting per second against the number of chunks the stream contains per second?

**Reply by getroot** [reply]
Different players and different conditions have different chunk request rates. This is unpredictable.

for example...

Many players request a very large number of chunks when starting out.
In the case of a specific version of VLC, it operates as Legacy HLS, and 3 or 4 segments are downloaded at the beginning. If the segment is 10 seconds long, it is requested only once every 30 seconds.
In the case of Safari, there are times when a lot of segments are downloaded under certain conditions (network speed? This information is not disclosed) in the beginning, and then partial segments are suddenly downloaded and followed.
Most players request a lot of segments in an instant when seeking.
If the network is unstable, it requests many segments from the past.
I remember some players (was it ffplay...?) every time they download a chunk, they disconnect and reconnect to download it.

Do you have to calculate the exact number of concurrent viewers? If so, have OvenPlayer and other supporting players send Play/Ping (periodic)/Stop events to your server. We do the same for projects that require accurate concurrent viewer counting. (If we can force users to only use the players we provide)

Counting the number of connected players at the server level is very difficult. Calculating by the number of TCP connections is actually very wrong when connecting with HTTP/1.1. (HTTP/1.1 connects 3-4 tcp connections per player. :-( ) 

This is similar in difficulty to an HTTP server counting the number of browsers connecting to it. At least I think the current algorithm is the most accurate. Except for the weak point where every time the user refreshes it gets +1... However, when actually operating the service, there were not that many users who refreshed compared to the total number of counts, so it was not a big problem. Until now, our customers have been more interested in throughput per stream than numbers. (since that's the cost)

**Reply by getroot** [reply]
I thought you wanted to count the number of connected players and answered it. But maybe what you really want is the number of connected devices (which is actually the number of users)? (Even if one person launches 10 players on one pc, it counts as 1)

**Reply by getroot** [reply]
There are also methods like this: At first, I also thought about the algorithm below and the current one.

- Assumes viewer is running 10 players on one PC
- 5 players playing app/stream/llhls.m3u8 -> assigned to session 1 -> counted as 1 concurrent viewer
- 5 players playing app/stream2/llhls.m3u8 -> assigned to session 2 -> counted as 1 concurrent viewer
- Calculate concurrent viewers by the total number of created sessions (currently 2)
- Does not change even after refreshing
- If there is no request for 30 seconds in session 1 -> delete session 1 (currently 1)

The downside of this is that it doesn't count one viewer playing one stream with multiple players.

Do you think this is better?

**Reply by bchah** [reply]
I definitely understand why people would mostly be interested in the total throughput. Here is one more idea then I will let this rest 😅

You mentioned that we increment the viewer count by 1 when the master playlist is requested. What if we only increment it one time per connection lifecycle per stream? That means User1 can only increment the LLHLS viewer count for Stream1 one time until they close the browser window or the TCP connection is otherwise terminated. It's a lot less mad science than my other idea and could prevent the "+1" issue. What do you think?

**Reply by getroot** [reply]
I asked a step faster than you. ;-)
https://github.com/AirenSoft/OvenMediaEngine/discussions/1313#discussioncomment-6468949

**Reply by bchah** [reply]
Replying to your edited comment: if my latest suggestion is no good, I think your updated suggestion is fantastic. 10 players of the same stream on the same screen is still one viewer in my opinion - and the throughout can still be measured for all 10 streams anyways so no info is lost. Interested to see what you think!

**Reply by getroot** [reply]
Your latest proposal (https://github.com/AirenSoft/OvenMediaEngine/discussions/1313#discussioncomment-6468965) and my last proposal (https://github.com/AirenSoft/OvenMediaEngine/discussions/1313#discussioncomment-6468949) essentially use the same algorithm.

I'm not sure which approach users would prefer between the current method and the one you suggested. You believe that calculating multiple players playing the same stream on one PC as one is a better approach? I fully respect your opinion. Please consider discussing this with your customers and let me know.

As I talk with you, I feel like there won't be many strange viewers who would watch the same stream on ten different players from one PC, so there may not be a need to consider them.

However, in the rare case that such peculiar viewers are very abundant, there could be disputes during cost settlement. The throughput would be very high, but the number of viewers might be significantly lower, potentially leading to disagreements when calculating expenses.

**Reply by bchah** [reply]
Of course the King of OME is faster than me 😂

You have shared so much important insight and information, I feel educated to rephrase my question now:

Would you consider adding a "uniqueViewerCount" metric to the Stream Stats API which uses the new calculation method we have discussed? 😁 Then the original method can remain too.

Thank you for all of your help and insight here @getroot 👑

**Reply by getroot** [reply]
If Meta is added, WebRTC or WHIP must also be considered, so let's think about the number of various cases.  :smiley:
First of all, I am positive about applying a new viewership calculation method to LLHLS.
